File: 787-Cheapest Flights Within K Stops.py
from collections import defaultdict, deque
class Solution(object):
    def findCheapestPrice(self, n, flights, src, dst, K):
        """
        :type n: int
        :type flights: List[List[int]]
        :type src: int
        :type dst: int
        :type K: int
        :rtype: int
        """
        graph = defaultdict(dict)
        for start, end, weight in flights:
            graph[start][end] = weight
            
        queue = deque()
        step = -1
        cost = 0
        queue.append((src, cost, step))
        ans = float('inf')
        
        while queue:
            for _ in range(len(queue)):
                cur, cost, step = queue.popleft()
                if step > K:
                    break
                if cur == dst:
                    ans = min(ans, cost)
                
                for nxt, weight in graph[cur].items():
                    if cost + weight < ans:
                        queue.append((nxt, cost+weight, step+1))
        
        return -1 if ans == float('inf') else ans

File: C++/787-Cheapest Flights Within K Stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        unordered_map<int, vector<pair<int, int>>> graph;
        
        for (auto flight:flights) {
            graph[flight[0]].push_back({flight[1], flight[2]});
        }
        
        priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, greater<tuple<int, int, int>>> pq;
        unordered_map<int, int> visited;
        for (int i=0; i<n; i++) {
            visited[i] = k+1;
        }
        
        pq.push({0, src, 0});
        
        while (!pq.empty()) {
            auto [tot, curr, step] = pq.top();
            pq.pop();
            if (step > visited[curr]) continue;
            visited[curr] = step;
            if (curr == dst) return tot;
            if (step > k) continue;
            
            for (auto [nxt, cost]:graph[curr]) {
                pq.push({tot+cost, nxt, step+1});
            }
        }
        
        return -1;
    }
};